File: problems-adapt/0470_flattest-route-across-a-grid/solution_dijkstra.ts
function flattestRoute(heights: number[][]): number {
    const rows = heights.length;
    const cols = heights[0].length;
    const INF = Infinity;
    // Bottleneck shortest path: Dijkstra with max in place of addition — a
    // path's effort is the largest height difference along it, and the
    // smallest tentative effort popped is already final.
    const dist: number[][] = Array.from({ length: rows }, () => new Array(cols).fill(INF));
    dist[0][0] = 0;

    // Min-heap of [d, r, c]
    const heap: number[][] = [];
    const push = (item: number[]): void => {
        heap.push(item);
        let i = heap.length - 1;
        while (i > 0) {
            const p = (i - 1) >> 1;
            if (heap[p][0] <= heap[i][0]) break;
            [heap[p], heap[i]] = [heap[i], heap[p]];
            i = p;
        }
    };
    const pop = (): number[] => {
        const top = heap[0];
        const last = heap.pop() as number[];
        if (heap.length > 0) {
            heap[0] = last;
            let i = 0;
            for (;;) {
                const l = 2 * i + 1;
                const r = l + 1;
                let m = i;
                if (l < heap.length && heap[l][0] < heap[m][0]) m = l;
                if (r < heap.length && heap[r][0] < heap[m][0]) m = r;
                if (m === i) break;
                [heap[m], heap[i]] = [heap[i], heap[m]];
                i = m;
            }
        }
        return top;
    };

    push([0, 0, 0]);
    const dirs: number[][] = [
        [1, 0],
        [-1, 0],
        [0, 1],
        [0, -1],
    ];
    while (heap.length > 0) {
        const [d, r, c] = pop();
        // The first time the goal is popped its effort is optimal.
        if (r === rows - 1 && c === cols - 1) {
            return d;
        }
        // Stale-entry guard: skip outdated heap records.
        if (d > dist[r][c]) {
            continue;
        }
        for (const [dr, dc] of dirs) {
            const nr = r + dr;
            const nc = c + dc;
            if (nr >= 0 && nr < rows && nc >= 0 && nc < cols) {
                const nd = Math.max(d, Math.abs(heights[nr][nc] - heights[r][c]));
                // Relax only when the bottleneck effort strictly improves.
                if (nd < dist[nr][nc]) {
                    dist[nr][nc] = nd;
                    push([nd, nr, nc]);
                }
            }
        }
    }
    return 0;
}

File: problems/1600-1700/1631_path-with-minimum-effort/solution_dijkstra.go
import "container/heap"

type effortHeap [][3]int

func (h effortHeap) Len() int            { return len(h) }
func (h effortHeap) Less(i, j int) bool  { return h[i][0] < h[j][0] }
func (h effortHeap) Swap(i, j int)       { h[i], h[j] = h[j], h[i] }
func (h *effortHeap) Push(x interface{}) { *h = append(*h, x.([3]int)) }
func (h *effortHeap) Pop() interface{} {
	old := *h
	n := len(old)
	item := old[n-1]
	*h = old[:n-1]
	return item
}

func minimumEffortPath(heights [][]int) int {
	rows := len(heights)
	cols := len(heights[0])
	const inf = int(^uint(0) >> 1)
	// Bottleneck shortest path: Dijkstra with max in place of addition — a
	// path's effort is the largest height difference along it, and the
	// smallest tentative effort popped is already final.
	dist := make([][]int, rows)
	for i := range dist {
		dist[i] = make([]int, cols)
		for j := range dist[i] {
			dist[i][j] = inf
		}
	}
	dist[0][0] = 0
	h := &effortHeap{}
	heap.Push(h, [3]int{0, 0, 0})
	dirs := [4][2]int{{1, 0}, {-1, 0}, {0, 1}, {0, -1}}
	for h.Len() > 0 {
		top := heap.Pop(h).([3]int)
		d, r, c := top[0], top[1], top[2]
		// The first time the goal is popped its effort is optimal.
		if r == rows-1 && c == cols-1 {
			return d
		}
		// Stale-entry guard: skip outdated heap records.
		if d > dist[r][c] {
			continue
		}
		for _, dir := range dirs {
			nr, nc := r+dir[0], c+dir[1]
			if nr >= 0 && nr < rows && nc >= 0 && nc < cols {
				nd := d
				diff := heights[nr][nc] - heights[r][c]
				if diff < 0 {
					diff = -diff
				}
				if diff > nd {
					nd = diff
				}
				// Relax only when the bottleneck effort strictly improves.
				if nd < dist[nr][nc] {
					dist[nr][nc] = nd
					heap.Push(h, [3]int{nd, nr, nc})
				}
			}
		}
	}
	return 0
}
